larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
+
+        for (unsigned int sliceIndex = 0; sliceIndex < nSlices; ++sliceIndex)
+        {
+            const SliceFeatures &features(sliceFeaturesVector.at(sliceIndex));
+            if (!features.IsFeatureVectorAvailable()) continue;
+
+            LArSvmHelper::ProduceTrainingExample(m_trainingOutputFile, sliceIndex == bestSliceIndex, features.GetFeatureVector());
+        }
+
+        return;
+    }
+
+    this->SelectPfosByProbability(nuSliceHypotheses, crSliceHypotheses, sliceFeaturesVector, selectedPfos);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::GetSliceFeatures(const NeutrinoIdTool *const pTool, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, SliceFeaturesVector &sliceFeaturesVector) const
+{
+    for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
+        sliceFeaturesVector.push_back(SliceFeatures(nuSliceHypotheses.at(sliceIndex), crSliceHypotheses.at(sliceIndex), pTool));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool NeutrinoIdTool::GetBestSliceIndex(const MasterAlgorithm *const pAlgorithm, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, unsigned int &bestSliceIndex) const
+{
+    unsigned int nHitsInBestSlice(0);
+    unsigned int nNuHitsInBestSlice(0);
+    unsigned int nuNHitsTotal(0);
+   
     for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
     {
-        const PfoList &sliceOutput((m_selectAllNeutrinos || (m_selectOnlyFirstSliceNeutrinos && (0 == sliceIndex))) ?
-            nuSliceHypotheses.at(sliceIndex) : crSliceHypotheses.at(sliceIndex));
+        if (nuSliceHypotheses.at(sliceIndex).size() != 1)
+            throw StatusCodeException(STATUS_CODE_OUT_OF_RANGE);
+
+        const PfoList &nuFinalStates(nuSliceHypotheses.at(sliceIndex).front()->GetDaughterPfoList());
 
-        selectedPfos.insert(selectedPfos.end(), sliceOutput.begin(), sliceOutput.end());
+        CaloHitList reconstructedHits;
+        this->Collect2DHits(nuFinalStates, reconstructedHits);
+        this->Collect2DHits(crSliceHypotheses.at(sliceIndex), reconstructedHits);
+
+        const unsigned int nNuHits(this->CountNeutrinoInducedHits(reconstructedHits));
+        nuNHitsTotal += nNuHits;
+        
+        // TODO may need to be careful with reproducibility here?
+        if (nNuHits > nNuHitsInBestSlice)
+        {
+            nNuHitsInBestSlice = nNuHits;
+            nHitsInBestSlice = reconstructedHits.size();
+            bestSliceIndex = sliceIndex; 
+        }
+    }
+
+    // ATTN for events with no neutrino induced hits, default neutrino purity and completeness to zero
+    float purity(0);
```

Comment:
Can use `? :` operator to collapse these down to two lines, and make them const floats. Usually set floats to e.g. `0.f` to make types explicit, rather than `0`